metricbeat/module/mssql/io/io.go

```diff
@@ -0,0 +1,104 @@
+package io
+
+import (
+	"database/sql"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/pkg/errors"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("mssql", "io", New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	counter int
+	db      *sql.DB
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The mssql io metricset is experimental.")
+
+	config := struct{}{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		counter:       1,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) (common.MapStr, error) {
+	return m.loadFileSpaceUsage(m.db)
+}
+
+func (m *MetricSet) loadFileSpaceUsage(db *sql.DB) (common.MapStr, error) {
+	// Returns the global status, also for versions previous 5.0.2
+	rows, err := db.Query("SELECT * FROM sys.dm_db_file_space_usage;")
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if closeErr := rows.Close(); closeErr != nil {
+			//TODO Log error? Ignore it?
+		}
+	}()
+
+	mapR, err := rowsToMapR(rows)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not convert row result to beats mapr")
+	}
+
+	return mapR, nil
+}
+
+func rowsToMapR(rows *sql.Rows) (common.MapStr, error) {
```

Comment:
We use to call this method `eventMapping` in other modules :slightly_smiling_face: